Feature: Forms

VincentFoulon80 opened the pull request:
I started implementing forms into `console_engine`, the objective is to allow the developer to easily create self-managed TUI forms.
They'll be enclosed in their own "form" feature. The "event" feature will also be needed in order for them to handle user input properly.

# 🆕 Features

- Forms that can handle multiple fields simultaneously
- Fields:
  - Text
  - HiddenText
  - Checkbox
  - Radio
- Basic Field Validation:
  - NotBlank
  - Integer
  - Number
  - Alphabetic
  - Alphanumeric
  - Callback
  - CharactersCallback

VincentFoulon80 commented:
This Pull Request is mostly finished, if anybody wants to review it, you're welcome !

LoipesMas commented on src/forms/text.rs:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    fn handle_event(&mut self, event: &Event) {
+        if self.active {
```
Personally I'd change this to:
```rust
if !self.active {
    return;
}
...
```
or alternatively only send events to those FormFields that are active.

LoipesMas commented on src/forms/form.rs:
```diff
@@ -0,0 +1,355 @@
+
+    /// Checks if the form is entirely valid. If any field fails its `validate` method, the function returns `false`
+    /// and a list of every reported error is stored in the form, waiting for [get_error](#methods.get_error) to be called
+    pub fn is_valid(&mut self) -> bool {
+        self.errors.clear();
+        for (name, field) in self.fields.iter() {
+            let mut field_errors: FormValidationResult = vec![];
+            field.validate(&mut field_errors);
+            if !field_errors.is_empty() {
+                self.errors.insert(String::from(name), field_errors);
+            }
+        }
+        let mut self_errors: FormValidationResult = vec![];
+        self.validate(&mut self_errors);
+
+        self.errors.is_empty() && self_errors.is_empty()
+    }
+}
```
This function feels weird to me.
We call `validate` on every field and we get their errors, but then we call `self.validate` which also calls `validate` on all the forms and gets the errors.
Also I'd suggest removing this function and just have `self.validate` return the errors and then user will check if errors is empty to see if form is valid. This would make storing errors and `get_errors` obsolete.
Thoughts?

VincentFoulon80 replied:
I may have been a bit tired when I worked on this, Forms are a special kind of field and I wanted them to be able to retrieve errors by field independently from each other. I'll try to rethink this part properly.

LoipesMas commented on src/forms/form.rs:
```diff
@@ -0,0 +1,355 @@
+    ///
+    /// You must run [is_valid](#methods.is_valid) first in order to be able to retrieve the errors
+    pub fn get_error(&self, name: &str) -> Option<&FormValidationResult> {
+        for (field_name, errors) in self.errors.iter() {
+            if name == *field_name {
+                return Some(errors);
+            }
+        }
+        None
```
Because `errors` is a hashmap we can instead do just `self.errors.get(name)`

LoipesMas commented on src/forms/form.rs:
```diff
@@ -0,0 +1,355 @@
+            self.screen.resize(self.screen.get_width(), total_height);
+        }
+        if self.dirty {
+            let padding = if self.options.style.border.is_some() {
+                1
+            } else {
+                0
+            };
```
In my opinion `let padding = self.options.style.border.is_some() as i32;` would look cleaner.

LoipesMas commented on src/forms/mod.rs:
```diff
@@ -0,0 +1,153 @@
+    /// Gets the options of the Field
+    fn get_options(&self) -> &FormOptions;
+    /// Tell if we should display the label externally (some fields may want to display it themselves like buttons)
+    fn display_label(&self) -> bool {
```
`display_label` sounds like an action (like `draw` or `validate`) but it doesn't actually display the label. I'd suggest renaming it, to something like `should_display_label` (or probably something else).

LoipesMas commented on src/forms/constraints/numbers.rs:
```diff
@@ -0,0 +1,147 @@
+            FormValue::Boolean(_) => true,
+            FormValue::Index(_) => true,
+            FormValue::String(value) => {
+                if let Some(chr) = value.chars().next() {
+                    if !chr.is_digit(10) && chr != '-' && chr != '+' {
+                        return false;
+                    }
+                }
+                value.chars().skip(1).all(|x| x.is_digit(10))
```
You could do `value.parse::<i32>().is_ok()` instead. But if you want to keep it like that it's also fine.

LoipesMas commented on src/forms/constraints/numbers.rs:
```diff
@@ -0,0 +1,147 @@
+            FormValue::Boolean(_) => true,
+            FormValue::Index(_) => true,
+            FormValue::String(value) => {
+                if let Some(chr) = value.chars().next() {
+                    if !chr.is_numeric() && chr != '-' && chr != '+' {
+                        return false;
+                    }
+                }
+                value
+                    .chars()
+                    .skip(1)
+                    .all(|x| x.is_numeric() || x == '.' || x == ',')
+                    && value.chars().filter(|&x| x == '.' || x == ',').count() <= 1
```
And here you could do `value.parse::<f32>().is_ok()`

VincentFoulon80 replied:
I wonder is parsing to a float accepts dots as well as commas for values, also do you think keeping it that way for the same reasons as for integers is acceptable ?

LoipesMas replied:
In my opinion, since it's a Rust library, validating it exactly like Rust does makes sense. Because now, if developer doesn't check the source code / documentation and tries to just `::parse<f32>` the value he can get an error even if the form is considered valid. (This also kinda applies to integer validation, but there isn't as big of a difference)
Also IIRC Rust also parses `3e10` etc., so you would also get that if you used standard Rust parsing.

VincentFoulon80 replied:
In this case parsing the input to a f64 is more relevant since floats allow values from -infinity to +infinity (aside from precision)

LoipesMas commented on src/forms/text.rs:
```diff
@@ -0,0 +1,387 @@
+        }
+        let current_cursor_pos =
+            std::cmp::min(self.cursor_pos as i32, self.screen.get_width() as i32 - 1);
+        if self.active && tick % 2 == 0 {
+            if let Ok(mut cursor_pxl) = self.screen.get_pxl(current_cursor_pos, 0) {
+                cursor_pxl.bg = self.options.style.fg;
+                cursor_pxl.fg = self.options.style.bg;
+                self.screen.set_pxl(current_cursor_pos, 0, cursor_pxl);
+            }
+        } else if let Ok(mut cursor_pxl) = self.screen.get_pxl(current_cursor_pos, 0) {
+            cursor_pxl.bg = self.options.style.bg;
+            cursor_pxl.fg = self.options.style.fg;
+            self.screen.set_pxl(current_cursor_pos, 0, cursor_pxl);
+        }
```
```rust
if let Ok(mut cursor_pxl) = self.screen.get_pxl(current_cursor_pos, 0) {
    if self.active && tick % 2 == 0 {
        cursor_pxl.bg = self.options.style.fg;
        cursor_pxl.fg = self.options.style.bg;
    }
    else {
        cursor_pxl.bg = self.options.style.bg;
        cursor_pxl.fg = self.options.style.fg;
    }
    self.screen.set_pxl(current_cursor_pos, 0, cursor_pxl);
}
```
I think this would be more readable.

LoipesMas reviewed (commented):
Overall it looks good :+1: 
One bigger change I'd suggest is "consuming" the events.
So if we have event handling `program -> form -> active field`, then if active field does something with the event (and somehow marks it as handled), then form won't process this further. And then also program wouldn't process it further.
This could be as simple as returning a bool from `handle_event` and returning early.
Not sure if there are any current problems that this would solve, but I think that it makes sense (and also I've seen it work like that it quite a few places).
What do you think?

LoipesMas commented:
>I'd like to allow the possibility of multiple fields handling a single event

This sounds like something that could potentially be a hard-to-find problem in the future (either in development of the crate or when using the crate). And I don't really see the use case for that. The possibility could maybe implemented in an explicit way, e.g. some UI element that sends events to all of it's children.

> we not only give the event to the form (so he may or may not use it) but after that we also use it afterward to handle for example form cancelling

Maybe better way to describe that was that first the program checks if it wants to do something with the event (quit, cancel, etc.) and if it decides that it doesn't then pass it forward. I think this makes more sense.
Then it could be implemented like this:
```rust
match event {
    Event::Key(KeyEvent { KeyCode::Escape, modifiers }) => {
        break;
    },
    event => form.handle_event(&event),
}
```
I'm just worried that this handling of events in multiple places could cause some unexpected behavior.

VincentFoulon80 commented:
Right, I see your point. I'll work on that part then, that shouldn't take too much time to implement thankfully.

LoipesMas commented:
One more thing: I wonder if it wouldn't be better if `get_result` would return a `Result` type that would have the validation error as `Err`. So instead of checking first if there are any errors and the proceeding with getting the form data, this would happen at once.
[Here's how I imagine it](https://gist.github.com/LoipesMas/5657778ff1f0993d3709db8ee51a5eb9) more or less.
This is more inline with Rust design (`Result` type was made for that). I used a separate function for getting the data from form so you can use `?` operator.

VincentFoulon80 commented:
> One more thing: I wonder if it wouldn't be better if `get_result` would return a `Result` type that would have the validation error as `Err`. So instead of checking first if there are any errors and the proceeding with getting the form data, this would happen at once. [Here's how I imagine it](https://gist.github.com/LoipesMas/5657778ff1f0993d3709db8ee51a5eb9) more or less. This is more inline with Rust design (`Result` type was made for that). I used a separate function for getting the data from form so you can use `?` operator.

That's genius! I didn't think of that as those two functions were made several days apart. I'll apply this once I'm back to it. Thanks for the suggestion !

LoipesMas commented on src/forms/text.rs:
```diff
@@ -0,0 +1,421 @@
+
+    /// Insert a character at the position of the cursor
+    pub fn put_char(&mut self, chr: char) {
+        let mut new_buffer = self
+            .input_buffer
+            .chars()
+            .take(self.cursor_pos)
+            .collect::<String>();
+        new_buffer.push(chr);
+        new_buffer.push_str(
+            &self
+                .input_buffer
+                .chars()
+                .skip(self.cursor_pos)
+                .collect::<String>(),
+        );
+        self.input_buffer = new_buffer;
```
Never mind, we can just `self.input_buffer.insert(self.cursor_pos, chr)`, right?

LoipesMas replied:
Actually, it uses byte offset (instead of char, as I would expect), so it can behave not as expected in some cases (or even panic when not on char boundary).

VincentFoulon80 replied:
Actually I changed thecode this way because of the byte offset instead of chars, I was getting panics when I wrote some special characters on the inputs

VincentFoulon80 replied:
Tested option 1, worked successfully, we don't even need that `.collect::<String>()`

I prefer this one as it gets straight to the point. I'm learning a lot with your suggestions 😄

LoipesMas replied:
>I'm learning a lot with your suggestions

That's great to hear! I'm learning too

LoipesMas commented on examples/form-choices.rs:
```diff
@@ -0,0 +1,127 @@
+
+    if form.is_finished() {
+        // Get the output of each fields
+        if let FormValue::Vec(selection_list) = form.get_result("checkbox").unwrap_or_default() {
```
I feel like using `unwrap_or_default` can potentially silence any errors that would occur.
Generally, I'm against using `unwrap_or_default` on `Result`s. On `Option`s it makes more sense.

VincentFoulon80 replied:
I agree with you on that point, it was a little side effect from changing the return value of `get_result`.

LoipesMas reviewed (commented):
I like the changes!

VincentFoulon80 commented on src/forms/mod.rs:
```diff
@@ -0,0 +1,159 @@
+    /// so the Validation Constraints provided by FormOptions will be automatically checked against the output of the field
+    fn validate(&self, validation_result: &mut FormValidationResult);
+    /// Get the output of the field
+    fn get_output(&self) -> FormValue;
```
I disagree, because Form also implement `get_output`, giving a HashMap of each of his fields.
`get_result_unvalidated`, on the other hand, returns an `Option<FormOutput>` that depends on the presence of a given field, also `get_result_unvalidated` takes an argument while `get_output` does not.

LoipesMas replied:
To me the name seems too different, since `get_result_unvalidated` is just a map to field's `get_output`. Maybe rename `get_result_unvalidated` to `get_field_output` or `get_output_of_field`?

VincentFoulon80 replied:
This might do the job, I named it that way due to its close relation with the get_result (it's more or less an error-free version of it)

LoipesMas replied:
I'm not sure which one would be better now. Maybe also rename `get_result` to `get_validated_field_output`?

VincentFoulon80 replied:
I prefer your suggestion on that, we keep the output name everywhere so no confusion about what we will get.
Tell me if it looks better for you

LoipesMas replied:
Yep, this looks better

VincentFoulon80 commented:
I addressed most of your comments,

I'm wondering something about validation constraints, precisely unhandled FormValue (like Nothing, Boolean, Index most of the time), should the behavior of constraints pass them through (return true even if they doesn't match the validation, to not block HashMap validation for example), or should they block (return false) ?

Starting this PR I was for the passing through, but by working on it right now I'm thinking about switching to a blocking behavior.

Any thoughts about this ?

LoipesMas commented on src/forms/constraints/logic.rs:
```diff
@@ -0,0 +1,152 @@
+impl FormConstraint for AllOf {
+    fn validate(&self, output: &FormValue) -> bool {
+        self.constraints.iter().all(|constraint| {
+            println!("{:?}", constraint.validate(output));
```
Left-over debugging print?

VincentFoulon80 replied:
😅

LoipesMas commented on src/forms/form.rs:
```diff
@@ -0,0 +1,393 @@
+
+    /// Get the (unvalidated) output of a specific field if it exists within the Form
+    pub fn get_result_unvalidated(&self, name: &str) -> Option<FormValue> {
+        for (field_name, field) in self.fields.iter() {
+            if name == *field_name {
```
Since you use this code in a few places, you could abstract it out to a separate function, e.g. `get_field_by_name(&self, name: &str) -> Option<_>`.
Also iterators have `.find` function, so maybe just using that would be enough to cut down the boilerplate code.
But abstracting it out would allow users to access fields.

Now I realize you do have `get_field` function, but it removes the field from the Form. To be fair I don't understand why.
Maybe rename it to `remove_field` and add `get_field` which doesn't remove the field?

VincentFoulon80 replied:
Actually, I'd like to do this, and that's what `get_field` would do .. I struggled (just a bit... a huge bit in fact 😓) to do it this way and the only solution I was able to make was this removing mess :/

I'll try to see if by any chance this works

LoipesMas replied:
I guess downcasting is the issue. I don't have much experience with it, so I don't have a solution ready, but I'll try to figure something out.

VincentFoulon80 replied:
I finally got it to work .. I don't know why it didn't work last time but I'll take it 👍

LoipesMas replied:
Nice!

LoipesMas commented:
> I addressed most of your comments,
> 
> I'm wondering something about validation constraints, precisely unhandled FormValue (like Nothing, Boolean, Index most of the time), should the behavior of constraints pass them through (return true even if they doesn't match the validation, to not block HashMap validation for example), or should they block (return false) ?
> 
> Starting this PR I was for the passing through, but by working on it right now I'm thinking about switching to a blocking behavior.
> 
> Any thoughts about this ?

I think it makes sense for them to be blocking, since the user decides what constraint are applied where. And being more picky could potentially prevent some unexpected behavior (e.g. applying number constraint to a toggling button doesn't make sense so it's better if it invalidates the form). 
Now I'm wondering if maybe it should be an error (rather than just not validating)? It fits Rust to be picky about types.

VincentFoulon80 commented:
> > I addressed most of your comments,
> > I'm wondering something about validation constraints, precisely unhandled FormValue (like Nothing, Boolean, Index most of the time), should the behavior of constraints pass them through (return true even if they doesn't match the validation, to not block HashMap validation for example), or should they block (return false) ?
> > Starting this PR I was for the passing through, but by working on it right now I'm thinking about switching to a blocking behavior.
> > Any thoughts about this ?
> 
> I think it makes sense for them to be blocking, since the user decides what constraint are applied where. And being more picky could potentially prevent some unexpected behavior (e.g. applying number constraint to a toggling button doesn't make sense so it's better if it invalidates the form). Now I'm wondering if maybe it should be an error (rather than just not validating)? It fits Rust to be picky about types.

I'd love to see compile time errors due to missuse of constraints but that implies that we know what kind of output a given field can return, and also what kind of output a given constraint is able to validate. But I have no idea if that's possible.

I'm not sure about runtime validation, should they return a Result or an Option (if not just a bool) ? I'm leaning toward Options but I'd like your opinion on this

LoipesMas commented:
Yeah, compile time errors are probably not feasible. But returning an Option or a Result doesn't seem like a good solution either, since then you have to decide whether you block on None/Err or pass. Or you could leave it up to the user. 
I was also thinking that maybe it could panic, but it feels kinda off too.

VincentFoulon80 commented:
Maybe with a 'support' function that returns bool if the given FormValue is supported ?

LoipesMas commented:
How would this work?

VincentFoulon80 commented:
Something around 
```rust
    fn supports(&self, output: &FormValue) -> bool {
        matches!(
            output,
            FormValue::Boolean(_) | FormValue::Index(_) | FormValue::List(_)
        )
    }
```
I'm not sure about this idea, it feels like bypassing something alike Options
I might simply return false for now (blocking), and figure out a better way later

LoipesMas commented:
Leaving it for later, when you have more experience with actually using it (and maybe feedback from users of the crate), is probably the right move right now.

VincentFoulon80 commented:
I think that's it,
By the way, I added a `IsTrue` constraint. I'll add some more later as I encounter some use-cases but I think I'll stop for this release. If you got any idea for more constraints I'll take them as well

LoipesMas commented:
I don't have any ideas for constraints for now

VincentFoulon80 commented:
Okay, then I think we're done, unless you got anything else to do before merging ?

LoipesMas commented:
Nope, looks good to me!

VincentFoulon80 commented:
So I'm merging this, thanks a million for your patience and all your reviews, the update looks amazing compared to my first attempts thanks to you.

By any chance, would you be interested in being a maintainer on this repository ?

LoipesMas commented:
Glad I could be of help!
I'm not familiar with or interested in TUI development so I don't think I'd contribute much, but feel free to mention me if you have any questions or want a code review and I'll try to help when I have the time.

VincentFoulon80 commented:
If you ever need help / review / opinion, feel free to tag me too, I would gladly try to help as well !
